app/controllers/settings/import_controller.rb

```diff
@@ -5,34 +5,35 @@ class Settings::ImportController < ApplicationController
   before_filter :authenticate_user!
 
   def myanimelist
-    hide_cover_image
+    params.require(:file)
 
-    if params[:animelist]
-      begin
-        if params[:animelist].content_type == "text/xml"
-          xml = params[:animelist].read
-        else
-          gz = Zlib::GzipReader.new(params[:animelist])
-          xml = gz.read
-          gz.close
-        end
+    # Prepare the file
+    # Check the magic number for gzip because some browsers are stupid
+    # Many users are uploading zipped-up lists with a text/xml MIME
+    if file.readpartial(3) == 0x1F_8B_08
+      gz = Zlib::GzipReader.new(file)
+      xml = gz.read
+      gz.close
+    elsif file.content_type.include?('xml')
+      xml = file.read
+    else
+      return error!(400, "Unknown format")
+    end
+    xml = XMLCleaner.clean(xml)
 
-        xml = XMLCleaner.clean(xml)
+    return error!(422, "Blank file") if xml.blank?
 
-        current_user.update_column :mal_import_in_progress, true
-        MyAnimeListImportApplyWorker.perform_async(current_user.id, xml)
+    # Queue the import
+    current_user.update_columns import_status: :queued,
+                                import_from: 'myanimelist'
+    MyAnimeListImportApplyWorker.perform_async(current_user.id, xml)
 
-        mixpanel.track "Imported from MyAnimeList", {email: current_user.email} if Rails.env.production?
-      rescue
-        flash[:error] = "There was a problem importing your anime list. Please send an email to <a href='mailto:vikhyat@hummingbird.me'>vikhyat@hummingbird.me</a> with the file you are trying to import.".html_safe
-        redirect_to "/users/edit"
-        return
-      end
-    end
+    render json: current_user
 
-    unless current_user.mal_import_in_progress?
-      flash[:success] = "Import completed successfully."
-      redirect_to user_path(current_user)
-    end
+    mixpanel.track "Imported from MyAnimeList", {email: current_user.email} if Rails.env.production?
+  rescue Exception
+    error! 500, "There was a problem importing your anime list.  Please send an
+                 email to vikhyat@hummingbird.me with the file you are trying
```

Comment:
Should we have like... a mailing list address for stuff like this?  `ninetynine-problems@hummingbird.me` or smth?